File: exercises/strain/.meta/src/reference/java/Strain.java
import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class Strain {

    public static <T> List<T> keep(List<T> collection, Predicate<T> predicate) {
        return filter(collection, predicate);
    }

    public static <T> List<T> discard(List<T> collection, Predicate<T> predicate) {
        return filter(collection, predicate.negate());
    }

    public static <T> List<T> filter(List<T> collection, Predicate<T> predicate) {
        List<T> filteredCollection = new ArrayList<>();

        for (T item : collection) {
            if (predicate.test(item)) {
                filteredCollection.add(item);
            }
        }

        return filteredCollection;
    }
}

File: go/strain/strain.go
package strain

type Lists []Ints

func (ls Lists) Keep(predicate func([]int) bool) Lists {
	var res Lists
	for _, list := range ls {
		if predicate(list) {
			res = append(res, list)
		}
	}
	return res
}

type Ints []int

func (l Ints) Keep(predicate func(x int) bool) Ints {
	var res Ints
	for _, elem := range l {
		if predicate(elem) {
			res = append(res, elem)
		}
	}
	return res
}

func (l Ints) Discard(predicate func(int) bool) Ints {
	var res Ints
	for _, elem := range l {
		if !predicate(elem) {
			res = append(res, elem)
		}
	}
	return res
}

type Strings []string

func (l Strings) Keep(predicate func(string) bool) Strings {
	var res Strings
	for _, elem := range l {
		if predicate(elem) {
			res = append(res, elem)
		}
	}
	return res
}
